yarn-project/pxe/src/database/kv_pxe_database.ts

```diff
@@ -214,6 +224,83 @@ export class KVPxeDatabase implements PxeDatabase {
+      for (const note of this.#notes.values()) {
+        const noteDao = IncomingNoteDao.fromBuffer(note);
+        if (noteDao.l2BlockNumber > blockNumber) {
+          const noteIndex = toBufferBE(noteDao.index, 32).toString('hex');
+          void this.#notes.delete(noteIndex);
+          void this.#nullifierToNoteId.delete(noteDao.siloedNullifier.toString());
+          for (const scope of this.#scopes.entries()) {
+            void this.#notesByAddressPointAndScope.get(scope)!.deleteValue(noteDao.addressPoint.toString(), noteIndex);
+            void this.#notesByTxHashAndScope.get(scope)!.deleteValue(noteDao.txHash.toString(), noteIndex);
+            void this.#notesByContractAndScope.get(scope)!.deleteValue(noteDao.contractAddress.toString(), noteIndex);
+            void this.#notesByStorageSlotAndScope.get(scope)!.deleteValue(noteDao.storageSlot.toString(), noteIndex);
+          }
+        }
+      }
+
+      for (const note of this.#outgoingNotes.values()) {
+        const noteDao = OutgoingNoteDao.fromBuffer(note);
+        if (noteDao.l2BlockNumber > blockNumber) {
+          const noteIndex = toBufferBE(noteDao.index, 32).toString('hex');
+          void this.#outgoingNotes.delete(noteIndex);
+          void this.#outgoingNotesByContract.deleteValue(noteDao.contractAddress.toString(), noteIndex);
+          void this.#outgoingNotesByStorageSlot.deleteValue(noteDao.storageSlot.toString(), noteIndex);
+          void this.#outgoingNotesByTxHash.deleteValue(noteDao.txHash.toString(), noteIndex);
+          void this.#outgoingNotesByOvpkM.deleteValue(noteDao.ovpkM.toString(), noteIndex);
+        }
+      }
+    });
+  }
+
+  public async unnullifyNotesAfter(blockNumber: number): Promise<void> {
+    const nullifiersToUndo: string[] = [];
+    let currentBlockNumber = blockNumber;
+    const maxBlockNumber = this.getBlockNumber() ?? currentBlockNumber;
+    while (currentBlockNumber <= maxBlockNumber) {
+      nullifiersToUndo.push(...this.#nullifiersByBlockNumber.getValues(currentBlockNumber));
+      currentBlockNumber++;
+    }
+
+    const notesIndexesToReinsert = await this.db.transaction(() =>
+      nullifiersToUndo.map(nullifier => this.#nullifiedNotesByNullifier.get(nullifier)),
+    );
+    const nullifiedNoteBuffers = await this.db.transaction(() => {
+      return notesIndexesToReinsert
+        .filter(noteIndex => noteIndex != undefined)
+        .map(noteIndex => this.#nullifiedNotes.get(noteIndex!));
+    });
+    const noteDaos = nullifiedNoteBuffers
+      .filter(buffer => buffer != undefined)
+      .map(buffer => IncomingNoteDao.fromBuffer(buffer!));
+
+    await this.db.transaction(() => {
+      for (const dao of noteDaos) {
+        const noteIndex = toBufferBE(dao.index, 32).toString('hex');
+        void this.#notes.set(noteIndex, dao.toBuffer());
+        void this.#nullifierToNoteId.set(dao.siloedNullifier.toString(), noteIndex);
+
+        // TODO: check if is this a good assumption
```

Comment:
I think we do store notes under their recipient's scope on automatic discovery, but notes can also be manually added under any scope, in which case this would be wrong since we'd not be restoring that correctly.

Reply:
Added scope tracking with yet another noteIndex -> whatever key value store